### PR TITLE
[Nano] Revise PyTorch Inference key feature doc regarding context manager

### PR DESCRIPTION
**Pull request by Oscilloscope98**
## Description

Revise PyTorch Inference key feature doc regarding context manager

### 1. Why the change?

https://github.com/analytics-zoo/nano/issues/267

### 2. Summary of the change 

- Revise regarding that all `InferenceOptimizer.optimize/trace/quantize` optimized models needs context manager for infernece
- Clarify for the multi-instance inference case
- Revise the installation note box 

### 4. How to test?
- [x] Document test: https://yuwentestdocs.readthedocs.io/en/nano-context-manager-doc-fix/doc/Nano/Overview/pytorch_inference.html

Revision on how-to guides will be raised in another PR.

### CODE REVIEW

**Review by MeouSker77 — APPROVED** [review]
LGTM

File: docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md [hunk 1]
```diff
@@ -41,6 +47,8 @@ Before you go ahead with these APIs, you have to make sure BigDL-Nano is correct
     - `ONNXRuntime <https://onnxruntime.ai/>`_: ``pip install onnx onnxruntime onnxruntime-extensions onnxsim neural-compressor``
 
     We recommand installing all dependencies by ``pip install --pre --upgrade bigdl-nano[pytorch,inference]``, because you may run into version issues if you install dependencies manually.
+
+    For the case you do not need the optimizations from INC, OpenVINO, or ONNXRuntime (e.g. you only want to use multi-instance inference, optimization from IPEX (Intel Extension for PyTorch) or TorchScript, etc.), you could just install BigDL-Nano for PyTorch through ``pip install --pre --upgrade bigdl-nano[pytorch]``.
```
Comment (TheaperDeng) [inline docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md:51]: From "We recommand ins..." to "--upgrade bigdl-nano[pytorch]``." is quite the same thing as the previous part of this note. Should we just delete them?

File: docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md [hunk 2]
```diff
@@ -26,13 +26,19 @@ Before you go ahead with these APIs, you have to make sure BigDL-Nano is correct
 .. note::
     You can install all required dependencies by
 
-    ::
+    .. code-block:: bash
 
         pip install --pre --upgrade bigdl-nano[pytorch,inference]
 
     This will install all dependencies required by BigDL-Nano PyTorch inference.
```
Comment (TheaperDeng) [inline docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md:33]: It's recommanded since it will install all dependencies required by BigDL-Nano PyTorch inference with no version conflict issue.

File: docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md [hunk 3]
```diff
@@ -138,7 +163,8 @@ Take the example in [Runtime Acceleration](#runtime-acceleration), and add quant
 ```python
 ov_q_model = InferenceOptimizer.quantize(model, accelerator='openvino', calib_data=dataloader)
 # run simple prediction with transparent acceleration
-y_hat = ov_q_model(x)
+with InferenceOptimizer.get_context(ort_q_model):
+    y_hat = ort_q_model(x)
```
Comment (TheaperDeng) [inline docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md:167]:
```python
with InferenceOptimizer.get_context(ov_q_model):
    y_hat = ov_q_model(x)
````

File: docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md [hunk 4]
```diff
@@ -326,15 +350,22 @@ multi_model = InferenceOptimizer.to_multi_instance(model, num_processes=4, cores
...
+```eval_rst
+.. note::
+    During multi-instance infernece, the context manager ``InferenceOptimizer.get_context(model=...)`` is not needed to be maunally added.
+```
+
 ## Automatic Context Management
-BigDL-Nano provides ``InferenceOptimizer.get_context(model=...)`` API to enable automatic context management for PyTorch inference. With only one line of code change, BigDL-Nano will automatically provide suitable context management for each accelerated model, it usually contains part of or all of following three types of context managers:
+BigDL-Nano provides ``InferenceOptimizer.get_context(model=...)`` API to enable automatic context management for PyTorch inference. With only one line of code change, BigDL-Nano will automatically provide suitable context management for each accelerated model optimized by ``InferenceOptimizer.trace``/``quantize``/``optimize``, it usually contains part of or all of following four types of context managers:
+
+1. ``torch.inference_mode(True)`` to disable gradients, which will be used for all models
```
Comment (TheaperDeng) [inline docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md:361]: For PT<=1.12, auto mixed precision inferencing will use `torch.no_grad()`

File: docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md (hunk 2 above)
Comment (Oscilloscope98) [reply docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md]: Thank you for the suggestion! I have updated the installation box as suggested.

File: docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md (hunk 3 above)
Comment (Oscilloscope98) [reply docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md]: Have fixed.

File: docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md (hunk 4 above)
Comment (Oscilloscope98) [reply docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md]: Have updated.

File: docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md [hunk 5]
```diff
@@ -326,15 +346,22 @@ multi_model = InferenceOptimizer.to_multi_instance(model, num_processes=4, cores
...
+```eval_rst
+.. note::
+    During multi-instance infernece, the context manager ``InferenceOptimizer.get_context(model=...)`` is not needed to be maunally added.
+```
+
 ## Automatic Context Management
-BigDL-Nano provides ``InferenceOptimizer.get_context(model=...)`` API to enable automatic context management for PyTorch inference. With only one line of code change, BigDL-Nano will automatically provide suitable context management for each accelerated model, it usually contains part of or all of following three types of context managers:
+BigDL-Nano provides ``InferenceOptimizer.get_context(model=...)`` API to enable automatic context management for PyTorch inference. With only one line of code change, BigDL-Nano will automatically provide suitable context management for each accelerated model optimized by ``InferenceOptimizer.trace``/``quantize``/``optimize``, it usually contains part of or all of following four types of context managers:
+
+1. ``torch.inference_mode(True)`` (when `torch` > 1.12) or ``torch.no_grad()`` (when `torch` <= 1.12) to disable gradients, which will be used for all models
```
Comment (rnwang04) [inline docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md:357]: There are still small problem here.
For PT<=1.12, only **auto mixed precision inferencing** will use torch.no_grad()
That it, for non-bf16 model, we will always use `torch.inference_mode(True)`.

Comment (Oscilloscope98) [reply docs/readthedocs/source/doc/Nano/Overview/pytorch_inference.md]: Has revised. Thank you!
